crates/ruff_python_semantic/src/context.rs

```diff
@@ -113,6 +114,55 @@ impl<'a> Context<'a> {
             .map_or(false, |binding| binding.kind.is_builtin())
     }
 
+    /// Resolve a reference to the given symbol.
+    pub fn resolve_reference(&mut self, symbol: &str, range: TextRange) -> ResolvedReference {
+        let mut first_iter = true;
+        let mut import_starred = false;
+        for scope_id in self.scopes.ancestor_ids(self.scope_id) {
+            let scope = &self.scopes[scope_id];
+            if scope.kind.is_class() {
+                if symbol == "__class__" {
+                    return ResolvedReference::ImplicitGlobal;
+                } else if !first_iter {
+                    continue;
+                }
+            }
+
+            if let Some(binding_id) = scope.get(symbol) {
+                // Mark the binding as used.
+                let context = self.execution_context();
+                self.bindings[*binding_id].mark_used(self.scope_id, range, context);
+
+                // But if it's a type annotation, don't treat it as resolved, unless we're in a
+                // forward reference. For example, given:
+                //
+                // ```python
+                // name: str
+                // print(name)
+                // ```
+                //
+                // The `name` in `print(name)` should be treated as unresolved, but the `name` in
+                // `name: str` should be treated as used.
+                if !self.in_deferred_type_definition()
+                    && self.bindings[*binding_id].kind.is_annotation()
```

Comment:
Nit: It feels a bit dangerous that the `resolve` behavior depends on where in the document we are. Meaning, it could return incorrect results if some rule does its own traversal and then calls `resolve`

Reply:
I don't like this. I also don't like that this method is mutative.